### PR TITLE
Support custom ellipsoids in Unity

### PR DESCRIPTION
**Pull request by azrogers**
In line with CesiumGS/cesium-unreal#1432 and CesiumGS/cesium-native#890, this change lets the Unity runtime support custom ellipsoids by creating ScriptableObjects for them and specifying those assets on CesiumGeoreference components.

### CODE REVIEW

**Comment by csciguy8** [comment]
This looks like a nice added feature. Can we get mention of it in CHANGES.md?

Also, it's not obvious to me what the typical workflow is to make a new custom ellipsoid, maybe that can be mentioned as well?

(I left this same comment on the cesium-unreal version of this PR too)

File: WGS84.asset
Comment (csciguy8) [inline WGS84.asset:1]: Verified this asset looks good in cesium-unity-samples, running with Unity 2021.3.10f1

**Comment by csciguy8** [comment]
Playing with the ellipsoids, found this odd side effect.

All I did was click on the Ellipsoid Override. Selected WGS84, then selected None. The terrain refreshed and had this weird squashed appearance.

image

File: Runtime/CesiumCartographicPolygon.cs
Comment (csciguy8) [inline Runtime/CesiumCartographicPolygon.cs:141]: `EllipsoidCenteredEllipsoidFixedToLongitudeLatitudeHeight` is really wordy.

Trying to think of a way to help this, would `CenteredFixedToLongitudeLatitudeHeight` by any better?

**Comment by csciguy8** [comment]
Found another odd behavior...

If I open up a sample scene, then switch the ellipsoid to WGS84, after the tileset refreshes, some tiles are missing. These tiles loaded just fine before I switched the ellipsoid...

image

File: native~/extern/cesium-native
Comment (csciguy8) [inline native~/extern/cesium-native:1]: It looks like we can update this all the way to June 20th (the last of the related ellipsoid commits).

This cesium-native update is only going to Jun 14th.

File: Runtime/CesiumSimplePlanarEllipsoidCurve.cs
Comment (csciguy8) [inline Runtime/CesiumSimplePlanarEllipsoidCurve.cs:25]: I'm guessing the mention of "Earth" in this function name is no longer valid?

File: Runtime/CesiumSimplePlanarEllipsoidCurve.cs
Comment (csciguy8) [inline Runtime/CesiumSimplePlanarEllipsoidCurve.cs:73]: I'm guessing the mention of "Earth" in this function name is no longer valid?

File: native~/Runtime/src/CameraManager.cpp
Comment (csciguy8) [inline native~/Runtime/src/CameraManager.cpp:77]: The nullptr check isn't needed here. `georeference` is a const reference, not a pointer.

File: native~/Runtime/src/UnityTilesetExternals.cpp
Comment (csciguy8) [inline native~/Runtime/src/UnityTilesetExternals.cpp:1]: I think you once had changes in this file, but ended up not needing them. Can revert all these.

We never actually store the ellipsoid in `TilesetExternals`

File: native~/Runtime/src/CesiumGlobeAnchorImpl.cpp
Comment (csciguy8) [inline native~/Runtime/src/CesiumGlobeAnchorImpl.cpp:196]: Can we call `getAnchorEllipsoid` just once here?

I'm assuming we are expecting the same ellipsoid between the first and second call.

File: native~/Runtime/src/CesiumBingMapsRasterOverlayImpl.cpp
Comment (csciguy8) [inline native~/Runtime/src/CesiumBingMapsRasterOverlayImpl.cpp:84]: Could replace this with `pTileset->getEllipsoid`, since `pTileset` has already been checked against null.

**Review by csciguy8 — CHANGES_REQUESTED** [review]
Thanks @azrogers! Looks like a lot of good changes to support the new functionality.

Requesting changes for some of the odd behavior that I tested.

Also a handful of code comments to consider too.

File: native~/Runtime/src/CameraManager.cpp
Comment (azrogers) [reply native~/Runtime/src/CameraManager.cpp:77]: Unity objects, including CesiumGeoreference, actually overload `!= nullptr` to work as an existence check. You can see it elsewhere in the file!

Comment (csciguy8) [reply native~/Runtime/src/CameraManager.cpp:77]: Ah!... not sure how I feel about that... but thanks for the explanation.

**Comment by azrogers** [comment]
There's a failing test with CesiumGlobeAnchor I'm still working on, but other than that I think I've updated everything.

**Comment by azrogers** [comment]
@csciguy8 should be ready for another look now!

**Review by csciguy8 — APPROVED** [review]
Looks great @azrogers !

Will merge once the final CI run completes.

Did encounter a minor refresh issue during testing. If I used a custom ellipsoid, then edited it, the tileset wouldn't refresh. I would have to manually click "Refresh Tileset". Not really sure if this is fixable, and doesn't seem substantial enough to block this PR.

File: native~/Runtime/src/CameraManager.cpp
Comment (kring) [reply native~/Runtime/src/CameraManager.cpp:77]: The way to think about it is that instances of the generated classes in the DotNet namespace act like C# references. And just like it makes sense to compare a C# reference to null, it makes sense to compare an instance of these wrapper classes to nullptr. It makes no sense to have a stack-allocated instance of a C# object because C# classes are - by definition - heap allocated. So an instance of these classes really can't be anything other than a reference.

Comment (csciguy8) [reply native~/Runtime/src/CameraManager.cpp:77]: It makes practical sense for sure. I think I was just caught a little off-guard since I was reviewing a .cpp file, and had my C++ glasses on.

**Pull request merged**
